src/notebooks/controllers/notebookControllerManager.ts

```diff
@@ -555,6 +608,78 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
             disposable.dispose();
         }
     }
+
+    // Use our kernel finder to rank our kernels, and see if we have an exact match
+    private async findPreferredKernelExactMatch(
+        document: NotebookDocument,
+        notebookMetadata: INotebookMetadata | undefined,
+        cancelToken: CancellationToken,
+        useCache: 'useCache' | 'ignoreCache' | undefined,
+        preferredInterpreter: PythonEnvironment | undefined
+    ): Promise<{
+        rankedConnections: KernelConnectionMetadata[] | undefined;
+        preferredConnection: KernelConnectionMetadata | undefined;
+    }> {
+        let preferredConnection: KernelConnectionMetadata | undefined;
+        const rankedConnections = await this.kernelFinder.rankKernels(
+            document.uri,
+            notebookMetadata,
+            cancelToken,
+            useCache
+        );
+
+        if (rankedConnections && rankedConnections.length) {
+            const potentialMatch = rankedConnections[rankedConnections.length - 1];
```

Comment:
seems really strange that the list is sorted worst to best.

Reply:
Probably because the 'best' used to have the 'highest' score. It would have taken reversing everything when Don wrote the ordering code.